### DISCUSSION TITLE
Announcing: easy Google Drive integration

### DISCUSSION THREAD

**Discussion by default-kramer** [TiddlyWiki/TiddlyWiki5#5660] [Ideas]
(Originally posted to Google Groups by mistake.)

I have prototyped a simple webpage that allows you to load a TiddlyWiki from your Google Drive and save the changes back. Perhaps the TiddlyWiki project will be interested in hosting a more polished version at my.tiddlywiki.com (for example)?

I built this so I can use my wiki from any browser that has internet access, as long as I can remember my Google credentials.

Website: https://default-kramer.github.io/
Source: https://github.com/default-kramer/default-kramer.github.io
How to Use It:
1. Create a fresh Google account if you don't trust me (and why should you?)
2. Create a folder named "asap.wiki" in the root of your Google Drive. (The name asap.wiki comes from an earlier idea of this project.)
3. Add exactly one TiddlyWiki file to the asap.wiki folder. It must be a new enough version of TiddlyWiki that includes the CustomSaver (added here https://github.com/Jermolene/TiddlyWiki5/pull/4908)
4. Browse to https://default-kramer.github.io/ and click the Authorize button. Continue past the "unverified app" warning.
5. The website should load the wiki from your Google Drive and display it in an iframe.
6. Make some changes and save your wiki.
7. Verify that the changes were written back to your Google Drive

Open Issues:
1. Obviously UI/UX is not my strong suit. Someone with more skill/experience than me could make this look and work a lot better.
2. If more than one wiki exists, it should prompt the user which one they want to open. (Right now it just fails.)
3. The app needs to be verified with Google. I think I am limited to 100 users right now because I am unverified.

Let me know if there is interest in making this an official TiddlyWiki offering and, if so, what I can do to help.

**Comment by pmario** [comment]
hi, There is a CLIENT_ID and API_KEY in the repo. Is it OK that those elements are public?

**Reply by default-kramer** [reply]
Thanks for looking, and yes, it is OK. Those keys are configured so that they can only be used from a certain host: in this case, default-kramer.github.io. Most of the code came from this tutorial: https://developers.google.com/drive/api/v3/quickstart/js

**Comment by pmario** [comment]
There should be some description, how I can host my own app like this. 

Did you try this mechanism with other "drive-like" providers?

**Reply by default-kramer** [reply]
If you wanted to host your own clone of this, you would need to create your own Google Developer account, set up your own CLIENT_ID and API_KEY per [these instructions](https://developers.google.com/drive/api/v3/quickstart/js), and then simply serve the HTML file with the new CLIENT_ID and API_KEY.

But the general public should not have to do this. Ideally, they just go my.tiddlywiki.com (or whatever) and start using it. Of course, some major UX improvements are needed before opening it up to the general public. If the TiddlyWiki project is interested, I am willing to make those improvements. But if not, I might just say "meh, it's good enough for me" and leave it as-is.

No, I have not tried any alternatives to Google Drive. I suspect that it is possible to support OneDrive and Dropbox.

**Comment by Jermolene** [comment]
That's great, thanks @default-kramer.

I wondered how it compares to https://github.com/tiddlydrive/tiddlydrive.github.io?

**Reply by default-kramer** [reply]
Oh nice, somehow I overlooked tiddlydrive.github.io. That looks to be exactly what I want.
